secretflow/device/device/teeu.py

```diff
@@ -141,6 +143,181 @@ def run(self, func: Callable, *args, **kwargs) -> TEEUData:
+                assert (
+                    value.device == self.device
+                ), f'unexpected device object {value.device} self {self.device}'
+                value_flat[i] = value.data
+        args, kwargs = jax.tree_util.tree_unflatten(value_tree, value_flat)
+
+        logging.debug(
+            (
+                f'Run method {name} of actor {self.actor_class}, num_returns='
+                f'{_num_returns}, args len: {len(args)}, kwargs len: {len(kwargs)}.'
+            )
+        )
+        handle = getattr(self.data, name)
+        res = handle.options(num_returns=_num_returns).remote(*args, **kwargs)
+        if _num_returns == 1:
+            return TEEUObject(self.device, res)
+        else:
+            return [TEEUObject(self.device, x) for x in res]
+
+    return wrapper
+
+
+def _cls_wrapper(cls):
+    def ray_get_wrapper(method):
+        def wrapper(self, *args, **kwargs):
+            logging.basicConfig(level=get_logging_level(), format=LOG_FORMAT)
+
+            logging.debug(
+                f'TEEU runs function: {method}, with args: {args}, kwargs: {kwargs}'
+            )
+
+            arg_flat, arg_tree = jax.tree_util.tree_flatten((args, kwargs))
+            refs = {
+                pos: arg
+                for pos, arg in enumerate(arg_flat)
+                if isinstance(arg, ray.ObjectRef)
+            }
+            actual_vals = ray.get(list(refs.values()))
+            for pos, actual_val in zip(refs.keys(), actual_vals):
+                arg_flat[pos] = actual_val
+
+            # Open the TEEUData.
+            teeu_data = [
+                (idx, value)
+                for idx, value in enumerate(arg_flat)
+                if isinstance(value, TEEUData)
+            ]
+
+            from secretflow.utils.cloudpickle import (
+                code_position_independent_dumps as dumps,
+            )
+
+            class_bytes = dumps(self.__class__.__bases__[0].__bases__[0], protocol=4)
+            data_keys = self.auth_frame.get_data_keys(
+                class_bytes, data_uuid_list=[o[1].data_uuid for o in teeu_data]
+            )
+
+            import ray.cloudpickle.cloudpickle as pickle
+            from cryptography.hazmat.primitives.ciphers.aead import AESGCM
+
+            for idx, value in teeu_data:
+                aesgcm = AESGCM(data_keys[idx])
+                new_value = aesgcm.decrypt(
+                    nonce=value.nonce, data=value.data, associated_data=value.aad
+                )
+                new_value = pickle.loads(new_value)
+                arg_flat[idx] = new_value
+
+            args_, kwargs_ = jax.tree_util.tree_unflatten(arg_tree, arg_flat)
+
+            return method(self, *args_, **kwargs_)
+
+        return wrapper
+
+    class ClassWithAuth(cls):
+        def __init__(self, *args, **kwargs):
+            from sdc.auth_frame import AuthFrame, CredentialsConf
+
+            auth_host = kwargs['auth_host']
```

Comment:
You can use `kwargs.pop('auth_host')` here so you don't need to delete it manually.